ports/espressif/common-hal/mdns/Server.c

```diff
@@ -0,0 +1,149 @@
+    common_hal_mdns_server_set_hostname(self, self->default_hostname);
+
+    // Set a delegated entry to ourselves. This allows us to respond to "circuitpython.local"
+    // queries as well.
+    // TODO: Allow for disabling this with `supervisor.disable_web_workflow()`.
+    mdns_ip_addr_t our_ip;
+    esp_netif_get_ip_info(common_hal_wifi_radio_obj.netif, &common_hal_wifi_radio_obj.ip_info);
+    our_ip.next = NULL;
+    our_ip.addr.type = ESP_IPADDR_TYPE_V4;
+    our_ip.addr.u_addr.ip4 = common_hal_wifi_radio_obj.ip_info.ip;
+    our_ip.addr.u_addr.ip6.addr[1] = 0;
+    our_ip.addr.u_addr.ip6.addr[2] = 0;
+    our_ip.addr.u_addr.ip6.addr[3] = 0;
+    our_ip.addr.u_addr.ip6.zone = 0;
+    mdns_delegate_hostname_add("circuitpython", &our_ip);
+}
+
+void common_hal_mdns_server_deinit(mdns_server_obj_t *self) {
+    inited = false;
+    mdns_free();
+}
+
+bool common_hal_mdns_server_deinited(mdns_server_obj_t *self) {
+    // This returns INVALID_STATE when not initialized and INVALID_PARAM when it
+    // is.
+    return mdns_instance_name_set(NULL) == ESP_ERR_INVALID_STATE;
+}
+
+const char * common_hal_mdns_server_get_hostname(mdns_server_obj_t *self) {
+    return self->hostname;
+}
+
+void common_hal_mdns_server_set_hostname(mdns_server_obj_t *self, const char *hostname) {
+    mdns_hostname_set(hostname);
+    self->hostname = hostname;
+}
+
+const char * common_hal_mdns_server_get_instance_name(mdns_server_obj_t *self) {
+    if (self->instance_name == NULL) {
+        return self->hostname;
+    }
+    return self->instance_name;
+}
+
+void common_hal_mdns_server_set_instance_name(mdns_server_obj_t *self, const char *instance_name) {
+    mdns_instance_name_set(instance_name);
+    self->instance_name = instance_name;
+}
+
+mp_obj_t common_hal_mdns_server_find(mdns_server_obj_t *self, const char *service_type, const char *protocol, mp_float_t timeout) {
+    mdns_search_once_t *search = mdns_query_async_new(NULL, service_type, protocol, MDNS_TYPE_PTR, timeout * 1000, 255, NULL);
+    if (search == NULL) {
+        mp_raise_RuntimeError(translate("Unable to start mDNS query"));
+    }
+    mdns_result_t *results;
+    while (!mdns_query_async_get_results(search, 1, &results)) {
+        RUN_BACKGROUND_TASKS;
+    }
+    mdns_query_async_delete(search);
+    // Count how many results we got.
+    // TODO: Remove this loop when moving off 4.4. Newer APIs will give us num_results
+    // back directly.
+    mdns_result_t *next = results;
+    uint8_t num_results = 0;
+    while (next != NULL) {
+        num_results++;
+        next = next->next;
+    }
+    mp_obj_tuple_t *tuple = MP_OBJ_TO_PTR(mp_obj_new_tuple(num_results, NULL));
+    next = results;
+    // Don't error if we're out of memory. Instead, truncate the tuple.
+    uint8_t added = 0;
+    while (next != NULL) {
+        mdns_remoteservice_obj_t *service = gc_alloc(sizeof(mdns_remoteservice_obj_t), true, false);
```

Comment:
Instead of `true` use `GC_ALLOC_FLAG_HAS_FINALISER`. Not using a finali(s/z)er here would be an easy mistake to make in a fresh implementation of Server; is there a way to move the allocation step to a place where it would be shared by all ports?

For that matter, why is RemoteService in the common-hal? If it's just a container for some data found during a search, then why not make it a kind of NamedTuple or similar, allowing it to live in shared-module.

Reply:
The need for a finaliser is implementation specific in this case (therefore common-hal) because this implementation is a wrapper around the IDF allocated mdns_result_t. We could move it to shared-module and not need a finaliser but that'd require a copy of all the data into the MP heap even if it is never read.

Reply:
If this trade-off needs to be changed in the future, we can do it. My gut would have said that since the CircuitPython heap is generally larger than the IDF heap it might be better to preemptively move it, especially since then we can rely on the GC to free it after it's no longer possible to use it.